File: benchmarks/tasks/pipeline/Pipeline.java
// Message passing: four producer threads each send fifty thousand numbers into
// one queue, and the main thread receives all of them and adds them up.

import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

public class Pipeline {
    static final int PRODUCERS = 4;
    static final int EACH = 50_000;

    public static void main(String[] args) throws InterruptedException {
        BlockingQueue<Long> q = new ArrayBlockingQueue<>(1024);
        for (int p = 0; p < PRODUCERS; p++) {
            final long base = (long) p * EACH;
            Thread t = new Thread(() -> {
                try {
                    for (int i = 0; i < EACH; i++) q.put(base + i);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            });
            t.setDaemon(true);
            t.start();
        }
        long total = 0;
        for (int n = 0; n < PRODUCERS * EACH; n++) total += q.take();
        System.out.println(total);
    }
}

File: benchmarks/tasks/pipeline/pipeline.py
"""Message passing: four producer threads each send fifty thousand numbers into
one queue, and the main thread receives all of them and adds them up.

`queue.Queue` is Python's channel. CPython's own lock means the producers take
turns rather than run at once, so this is the cost of the queue and not of
parallelism."""

import queue
import threading

PRODUCERS = 4
EACH = 50_000


def produce(q, p):
    for i in range(EACH):
        q.put(p * EACH + i)


def main():
    q = queue.Queue(maxsize=1024)
    for p in range(PRODUCERS):
        threading.Thread(target=produce, args=(q, p), daemon=True).start()
    total = 0
    for _ in range(PRODUCERS * EACH):
        total += q.get()
    print(total)


if __name__ == "__main__":
    main()
